File: examples/go/tests/json_schema_kb_test.go
package tests

import (
	"testing"

	apikb "examples/go/json_schema/api/kb"

	"github.com/stretchr/testify/require"
	"go.temporal.io/sdk/converter"
	"go.temporal.io/sdk/workflow"
)

// TestJSONSchemaKBRuntime round-trips every KB wire fixture through the Temporal
// default data converter and asserts JSON-equality against the canonical
// fixtures, mirroring the Python and Java suites.
//
// Exception (see json-schema/nullability.md): optional+nullable fields collapse
// in Go — an explicit wire `null` on such a field round-trips to absent. The
// fixtures carrying such nulls (page.json and block.json both hold `page: null`,
// which is optional+nullable) are therefore verified by deserialization + field
// checks rather than exact JSON-equality, matching the Java test.
func TestJSONSchemaKBRuntime(t *testing.T) {
	dc := converter.GetDefaultDataConverter()

	// page.json carries a nested block with page: null (optional+nullable),
	// which Go collapses on serialize, so verify by deserialization only.
	page := decodeFixture[apikb.Page](t, dc, "kb", "page.json")
	require.Equal(t, "page-1", page.PageId)
	require.Equal(t, "nex-gen", page.Meta.Author)
	require.Len(t, page.Blocks, 1)
	require.Equal(t, "block-1", page.Blocks[0].BlockId)
	require.Nil(t, page.Blocks[0].Page)
	require.NotNil(t, page.Blocks[0].Style)
	require.NotNil(t, page.Blocks[0].Style.Bold)
	require.True(t, *page.Blocks[0].Style.Bold)

	// block.json carries page: null (optional+nullable) — deserialization only.
	block := decodeFixture[apikb.Block](t, dc, "kb", "block.json")
	require.Equal(t, "block-1", block.BlockId)
	require.Equal(t, int64(0), block.Order)
	require.Nil(t, block.Page)
	require.NotNil(t, block.Style)
	require.NotNil(t, block.Style.Bold)
	require.True(t, *block.Style.Bold)

	category := roundTripJSONEq[apikb.Category](t, dc, "kb", "category-tree.json")
	require.Equal(t, "root", category.Id)
	require.Len(t, category.Children, 1)
	require.Equal(t, "child", category.Children[0].Id)

	getPage := roundTripJSONEq[apikb.GetPageInput](t, dc, "kb", "get-page-input.json")
	require.Equal(t, "page-1", getPage.PageId)

	getTree := roundTripJSONEq[apikb.GetCategoryTreeInput](t, dc, "kb", "get-category-tree-input.json")
	require.Equal(t, "root", getTree.RootId)

	putBlock := roundTripJSONEq[apikb.PutBlockOutput](t, dc, "kb", "put-block-output.json")
	require.Equal(t, "block-1", putBlock.BlockId)
	require.Equal(t, int64(7), putBlock.Revision)

	require.Equal(t, "example.kb.v1.KnowledgeBaseService", apikb.KnowledgeBaseService.ServiceName)
	require.Equal(t, "GetPage", apikb.KnowledgeBaseService.GetPage.Name())
	require.Equal(t, "PutBlock", apikb.KnowledgeBaseService.PutBlock.Name())
	require.Equal(t, "GetCategoryTree", apikb.KnowledgeBaseService.GetCategoryTree.Name())
	require.NotNil(t, apikb.NewKnowledgeBaseServiceClient("kb-endpoint"))
}

var _ func(*apikb.KnowledgeBaseServiceClient, workflow.Context, apikb.GetPageInput) workflow.Future = (*apikb.KnowledgeBaseServiceClient).GetPage
var _ func(*apikb.KnowledgeBaseServiceClient, workflow.Context, apikb.Block) workflow.Future = (*apikb.KnowledgeBaseServiceClient).PutBlock
var _ func(*apikb.KnowledgeBaseServiceClient, workflow.Context, apikb.GetCategoryTreeInput) workflow.Future = (*apikb.KnowledgeBaseServiceClient).GetCategoryTree

File: examples/typescript/tests/json-schema-kb.test.ts
import { fileURLToPath } from "node:url";
import { describe, expect, test } from "vitest";
import * as nexus from "nexus-rpc";

import {
  BlockMapper,
  CategoryMapper,
  GetCategoryTreeInputMapper,
  GetPageInputMapper,
  PageMapper,
  PutBlockOutputMapper,
} from "../json_schema/definitions/kb/index.ts";
import {
  knowledgeBaseService,
  type Category,
  type Page,
  type PutBlockOutput,
} from "../json_schema/api/kb/index.ts";
import { executeWorkflowWithNexus, withWorkflowEnvironment } from "./helpers.ts";
import {
  fixtureBytes,
  loadFixture as loadFixtureFrom,
  roundTripFixture,
  type IntermediateMapper,
} from "./json-converter-helper.ts";

const wireFixtureDir = new URL("../../wire/json_schema/kb/", import.meta.url);
const workflowsPath = fileURLToPath(
  new URL("./workflows/json-schema-kb.ts", import.meta.url),
);

function loadFixture<T = unknown>(name: string): T {
  return loadFixtureFrom<T>(wireFixtureDir, name);
}

// Round-trip a fixture through the Temporal data converter (driven by the
// generated mapper) and assert the re-serialized JSON is JSON-equal to the
// fixture. TS mappers preserve explicit nulls, so all KB fixtures use exact
// JSON-equality (no optional+nullable collapse — unlike Go).
function expectRoundTrip<T>(name: string, mapper: IntermediateMapper<T>): T {
  const { value, serialized } = roundTripFixture(
    mapper,
    fixtureBytes(wireFixtureDir, name),
  );
  expect(serialized).toEqual(loadFixture(name));
  return value;
}

describe("json-schema KB generated output", () => {
  test("roundtrips multi-file KB fixtures through the Temporal converter", () => {
    const page = expectRoundTrip("page.json", new PageMapper());
    expect(page.pageId).toBe("page-1");
    expect(page.blocks?.[0]?.blockId).toBe("block-1");
    expect(page.blocks?.[0]?.page).toBeNull();
    expect(page.blocks?.[0]?.style?.bold).toBe(true);

    const block = expectRoundTrip("block.json", new BlockMapper());
    expect(block.blockId).toBe("block-1");
    expect(block.page).toBeNull();

    const category = expectRoundTrip("category-tree.json", new CategoryMapper());
    expect(category.children?.[0]?.id).toBe("child");

    const request = expectRoundTrip("get-page-input.json", new GetPageInputMapper());
    expect(request.pageId).toBe("page-1");

    const categoryRequest = expectRoundTrip(
      "get-category-tree-input.json",
      new GetCategoryTreeInputMapper(),
    );
    expect(categoryRequest.rootId).toBe("root");

    const response = expectRoundTrip(
      "put-block-output.json",
      new PutBlockOutputMapper(),
    );
    expect(response.revision).toBe(7);
  });

  test("uses all generated KB operations through a real Nexus client", async () => {
    await withWorkflowEnvironment(async (env) => {
      const calls: Array<[string, unknown]> = [];
      const handler = nexus.serviceHandler(knowledgeBaseService, {
        async getPage(_ctx, input) {
          calls.push(["GetPage", input]);
          expect(input).toEqual({ pageId: "page-1" });
          return loadFixture<Page>("page.json");
        },
        async putBlock(_ctx, input) {
          calls.push(["PutBlock", input]);
          expect(input).toMatchObject({
            blockId: "block-1",
            order: 0,
            style: { bold: true, indent: 1 },
          });
          return loadFixture<PutBlockOutput>("put-block-output.json");
        },
        async getCategoryTree(_ctx, input) {
          calls.push(["GetCategoryTree", input]);
          expect(input).toEqual({ rootId: "root" });
          return loadFixture<Category>("category-tree.json");
        },
      });

      const result = await executeWorkflowWithNexus<{
        blockId: string;
        categoryChildId: string | undefined;
        pageId: string;
        revision: number;
      }>(env, {
        endpoint: "knowledge-base",
        nexusServices: [handler],
        workflowType: "jsonSchemaKbCaller",
        workflowsPath,
      });

      expect(result).toEqual({
        blockId: "block-1",
        categoryChildId: "child",
        pageId: "page-1",
        revision: 7,
      });
      expect(calls.map(([operation]) => operation)).toEqual([
        "GetPage",
        "PutBlock",
        "GetCategoryTree",
      ]);
    });
  });
});
